core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -174,22 +179,56 @@ class GroupCoordinator(val brokerId: Int,
       } else {
         val newMemberId = clientId + "-" + group.generateMemberIdSuffix
 
-        if (requireKnownMemberId) {
-          // If member id required, register the member in the pending member list
-          // and send back a response to call for another join group request with allocated member id.
-          group.addPendingMember(newMemberId)
-          addPendingMemberExpiration(group, newMemberId, sessionTimeoutMs)
-          responseCallback(joinError(newMemberId, Errors.MEMBER_ID_REQUIRED))
+        if (group.hasStaticMember(groupInstanceId)) {
+          val oldMemberId = group.getStaticMemberId(groupInstanceId)
+
+          if (group.is(Stable)) {
+            debug(s"Static member $groupInstanceId with unknown member id rejoins, assigning new member id $newMemberId. " +
+              s"No rebalance will be triggered.")
+
+            val oldMember = group.replace(oldMemberId, newMemberId, groupInstanceId)
+
+            // Heartbeat of old member id will expire without affection since the group no longer contains that member id.
+            // New member id will be the same.
+            completeAndScheduleNextHeartbeatExpiration(group, oldMember)
+
+            // For leader rejoin, we have to send back the entire group metadata, otherwise assignor will trigger
```

Comment:
Can you elaborate on this comment? I'm not sure I understand the problem.

Reply:
I don't think that comment is needed here. It's just an edge I caught during my experiment.